tensilelite/Tensile/SolutionStructs.py

```diff
@@ -1956,6 +1956,26 @@ def isDirectToVgprDoable(state, tc):
       reject(state, "DirectToVgpr%c does not supports Sparse"%(tc))
       return False
 
+    # for DTVA, does not work with PGR0
+    if tc == 'A' and state["PrefetchGlobalRead"] == 0:
+      reject(state, "DirectToVgpr%c does not supports PrefetchGlobalRead == 0."%(tc))
+      return False
+    
+    # for DTVA, does not work with NN and TLDS0
+    if tc == 'A' and state["TransposeLDS"] == 0 and (not state["ProblemType"]["TransposeA"] and not state["ProblemType"]["TransposeB"]):
+      reject(state, "DirectToVgpr%c does not supports NN case with TransposeLDS == 0."%(tc))
+      return False
+
+    # for DTVA, does not work with TT and Tail-loop
+    if tc == 'A' and (state["ProblemType"]["TransposeA"] and state["ProblemType"]["TransposeB"]):
+        reject(state, "DirectToVgpr%c does not supports TT case with Tail Loop."%(tc))
+        return False
+
+    # for DTVB, does not work with NN and Tail-loop
+    if  tc == 'B' and (not state["ProblemType"]["TransposeA"] and not state["ProblemType"]["TransposeB"]):
+        reject(state, "DirectToVgpr%c does not supports NN cases with Tail Loop."%(tc))
+        return False
+    
```

Comment:
I don't suggest this, since this blocks all TT, and NN not just tail-loop.
In my opinion, you should change the following part of Contractions.py:
in CompoundPredicates()

```
        if state['ProblemType']['SwizzleTensorA']:
            rv += [cls('SwizzleTensorA', value=state['ProblemType']['SwizzleTensorA'])]
            # TODO- (TT + DTVA) tail-loop is not working yet.
            if state['ProblemType']['TransposeB']:
                rv += [cls("BoundSizeMultiple", index=-1, value=state['DepthU'])]
```

Moving the #TODO part out from swizzle, and put to a block like:
 `if DTVA and TT, then rv += BoundSizeMultiple....`
And add the same block for `if DTVB and NN, then rv += BoundSizeMultiple`
Since we are "blocking the incorrect tail-loop case", not the "entire problem-type"

Reply:
It shoudn't directly set BoundSizeMultiple here, but should reset AssertSummationElementMultiple as DepthU in SolutionStruct.py if DepthU is larger than AssertSummationElementMultiple  for required conditions.
Ex:
if <condition satisfied>:
  AssertSummationElementMultiple  = max(AssertSummationElementMultiple, DepthU)

Reply:
As long as I tried before, some optimizations with AssertSummationElementMultiple = DepthU had some issues (even before I implemented DTVA) and I could not use it.
Besides that, I was not sure if AssertSummationElementMultiple = DepthU works with GSU.
In Tensile, NoTailLoop is enabled if AssertSummationElementMultiple is multiple of DepthU/GSU.

Reply:
Oh, I see what @jichangjichang means, if we force set AssertSummation, then in Contractions.py, ProblemPredicates . FromOriginalKeyPair() will automatically add that predicate instead of directly set it. As long as we can make sure the tail-loop size can be part of the predicates then I think it is good.

Reply:
(typo in my previous comment... )
AssertSummationElementMultiple is multiple of DepthU/GSU.
->AssertSummationElementMultiple is multiple of DepthU * GSU.

OK.
As long as you can confirm ASEM works with GSU, I am OK.